drivers/ieee802154/ieee802154_kw41z.c

```diff
@@ -24,12 +24,133 @@
+ * can have a breakpoint in the __reserved function (reset) and when
+ * the breakpoint happens look at the contents of the mtb buffer at
+ * the addresses traced.
+ *
+ * In gdb use: 'info line *<addr>'
+ *
+ */
+/*#define KW41Z_DEBUG_MTB_ENABLED   1*/
+
+#if KW41Z_DEBUG_MTB_ENABLED
+/*
+ * Stores 2 ^ TRAC_BUFFER_MAGNITUDE_PACKETS.
+ * 7 -> 128 packets
+ */
+#define TRACE_BUFFER_MAGNITUDE_PACKETS  7
+#define TRACE_BUFFER_SIZE (1 << (TRACE_BUFFER_MAGNITUDE_PACKETS + 1))
+
+__aligned__(TRACE_BUFFER_SIZE * sizeof(u32_t)) u32_t mtb[TRACE_BUFFER_SIZE];
+#endif
+
+/*
+ * For non-invasive tracing of IRQ events. Sometimes the print logs
+ * will shift the timings around so this trace buffer can be used to
+ * post inspect conditions to see what sequence of events occurred.
+ */
+/*#define KW41_DBG_TRACE_ENABLE	1*/
+
+#define KW41_DBG_TRACE_WTRM	0
+#define KW41_DBG_TRACE_RX	1
+#define KW41_DBG_TRACE_TX	2
+#define KW41_DBG_TRACE_CCA	3
+#define KW41_DBG_TRACE_TMR3	0xFF
+
+#ifdef KW41_DBG_TRACE_ENABLE
+
+#define KW41_DBG_TRACE_SIZE 30
+
+struct kw41_dbg_trace {
+	u8_t	type;
+	u32_t	irqsts;
+	u32_t	phy_ctrl;
+	u32_t	seq_state;
+};
+
+struct kw41_dbg_trace kw41_dbg[KW41_DBG_TRACE_SIZE];
+int kw41_dbg_idx;
+
+#define KW_DBG_TRACE(_type, _irqsts, _phy_ctrl, _seq_state) do { \
+		kw41_dbg[kw41_dbg_idx].type = (_type); \
+		kw41_dbg[kw41_dbg_idx].irqsts = (_irqsts); \
+		kw41_dbg[kw41_dbg_idx].phy_ctrl = (_phy_ctrl); \
+		kw41_dbg[kw41_dbg_idx].seq_state = (_seq_state); \
+		if (++kw41_dbg_idx == KW41_DBG_TRACE_SIZE) { \
+			kw41_dbg_idx = 0; \
+		} \
+	} while (0)
+
+#else
+
+#define KW_DBG_TRACE(_type, _irqsts, _phy_ctrl, _seq_state)
+
+#endif
+
+
+
+/*
+ * Frame Control Field and sequence number
+ * See Section 5.2.1.1
+ * TODO: Need to move the ieee802154 definition of this to the /include
+ * directory.
+ */
+struct ieee802154_fcf_seq {
+	struct {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+		u16_t frame_type		:3;
+		u16_t security_enabled		:1;
+		u16_t frame_pending		:1;
+		u16_t ar			:1;
+		u16_t pan_id_comp		:1;
+		u16_t reserved			:1;
+		u16_t seq_num_suppr		:1;
+		u16_t ie_list			:1;
+		u16_t dst_addr_mode		:2;
+		u16_t frame_version		:2;
+		u16_t src_addr_mode		:2;
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+		u16_t reserved			:1;
+		u16_t pan_id_comp		:1;
+		u16_t ar			:1;
+		u16_t frame_pending		:1;
+		u16_t security_enabled		:1;
+		u16_t frame_type		:3;
+		u16_t src_addr_mode		:2;
+		u16_t frame_version		:2;
+		u16_t dst_addr_mode		:2;
+		u16_t ie_list			:1;
+		u16_t seq_num_suppr		:1;
+#endif
+	} fc __packed;
+
+	u8_t sequence;
+} __packed;
+
+#if CONFIG_SYS_LOG_IEEE802154_DRIVER_LEVEL == SYS_LOG_LEVEL_DEBUG
+
+static void dump_802154(u8_t *payload)
+{
+	struct ieee802154_fcf_seq *mpdu = (struct ieee802154_fcf_seq *)payload;
+
+	SYS_LOG_DBG("FC: %d, ack: %d, seq: %d",
+		(payload[0])|(payload[1]<<8), mpdu->fc.ar, mpdu->sequence);
+}
+
+#else
+#define dump_802154(payload)
+#endif
```

Comment:
I don't like the idea of dumping 15.4 packet parts in the driver itself. Usually, knowing that frame is valid or not - on driver level - is straight forward.  Rest is L2's job.

Reply:
This is simply debug support when debugging just the driver. I need to keep the noise in the log down and wanted to see what packet was going out and then correlate it to wireshark sniffs

Reply:
Yep, so you don't need that. If you want to see packet internals, that's all about L2's job.
See CONFIG_NET_DEBUG_L2_IEEE802154_DISPLAY_PACKET

Reply:
Sorry, I was aware of the debug display stuff in L2 but turning it on put out too much information and at multiple points in the stack that wasn't as useful for the driver (useful for the L2 debug I'm sure). What I needed was something much closer to the HW so I could correlate what was physically being sent out the RF at the driver level with what wireshark was showing me. I didn't need the whole packet as I was just printing a single line with the FC, AR, and SEQ. 

The specific case this was useful is that wireshark was showing me that every frame was being retried and I didn't immediately understand why that was. Turned out that the L2 was sending frames on a retry because the driver was returning an error on each transmit. I realized this was happening when I added this particular packet dump. I found this much more useful for the driver debug needs.

Reply:
Please remove this. I don't want to see part of L2 being copied in drivers.

If you want a more flexible packet output in L2, let's make it in L2 (more Kconfig options to requests only the input, or the output, or both, only the FCF or all etc...)

Reply:
Can you check PR #4746 ? It's a first step towards getting more flexible packet display, in a generic way so all drivers will benefit from it. Thus you could remove this ;)